### ISSUE TITLE
What is the role of the manager "fabric.no_backward_sync" in lora.py?

### ISSUE THREAD

**Issue by richardsun-voyager** [Lightning-AI/lit-gpt#676]
I tried to finetune LLama2-13B with "finetune/lora.py" on my A100 server (2 GPUs, 40G each), but it always popped out OOM. 
Then I masked the manager " with fabric.no_backward_sync(model, enabled=is_accumulating):" and found it could work without OOM errors.
My question: what is the role of this manager? Do we really need it? It does not exist in the original Lit-LLama project.
Thanks!

**Comment by carmocca** [comment]
See https://lightning.ai/docs/fabric/stable/api/fabric_methods.html#no-backward-sync

Are you saying that you OOM only if it's used? I assume you are using FSDP, correct?

**Comment by richardsun-voyager** [comment]
Hi carmocca! Yes, OOM appears when it (" with fabric.no_backward_sync(model, enabled=is_accumulating):" ) is used, which looks weird to me. And in "lora.py", with two GPUs, the strategy will be FSDP. Are there any methods to solve this OOM issue even with this context manager on?

**Comment by carmocca** [comment]
I didnt know this, but it is expected.
It's mentioned in a note in https://pytorch.org/docs/stable/fsdp.html#torch.distributed.fsdp.FullyShardedDataParallel.no_sync

If you remove it, make sure that you also disable gradient accumulation, as it also mentions:

> FSDP currently does not support gradient accumulation outside no_sync() when using CPU offloading. Trying to do so yields incorrect results since FSDP will use the newly-reduced gradient instead of accumulating with any existing gradient.

I'm thinking that the Trainer could raise an error about this accumulation issue at least.

@awaelchli fyi

**Comment by richardsun-voyager** [comment]
I see. Thanks a lot!

**Comment by awaelchli** [comment]
I'd like to keep the context manager in the scripts please. There would be a substantial slowdown if it was removed!

> I'm thinking that the Trainer could raise an error about this accumulation issue at least.

But then you are suggesting that the user can't do gradient accumulation with FSDP? Can you elaborate?

**Comment by carmocca** [comment]
`if using_fsdp and gradient_accumulation_enabled and using_no_sync: raise NotSupported`

**Comment by awaelchli** [comment]
But why? This would incur a huge perf impact. I am definitely against this. Right now, IMO we are doing everything right in the Trainer.
